### PR TITLE
feat: plugin loader

### PR DESCRIPTION
**Pull request by arendjr**
## Summary

Continuation of https://github.com/biomejs/biome/pull/4071

Note this doesn't integrate the plugins into the analyzer yet, like the previous one did. This just makes sure they load as we want them to.

## Test Plan

Tests added. CI should remain green.

### CODE REVIEW

**Comment by ematipico** [comment]
Should we change the destination of the PR to `next`?

**Comment by arendjr** [comment]
I've rebased onto `next`. I've also included the Grit upgrade, because it wouldn't cleanly apply otherwise.

**Comment by arendjr** [comment]
Btw, thinking of the whole manifest thing, while I like this approach for future compatibility and the ability to publish plugins, I think I’ll still implement the ability to refer to refer to Grit files directly from the `plugins` array. It should be easy enough for us to detect and then instantiate a single-rule plugin and it’ll greatly simplify the use case when someone wants to have one or a few rules in their own repository.

File: crates/biome_plugin_loader/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+[package]
```
Comment (ematipico) [inline crates/biome_plugin_loader/Cargo.toml:1]: It's possible you created the crate manually or with cargo. If so you, you need to update manually `knope.toml` and add the crate. We have `just new-crate` that updates the file automatically

File: crates/biome_plugin_loader/src/analyzer_grit_plugin.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                None::<TextRange>,
+                markup!(<Emphasis>{name}</Emphasis>" errored: "<Error>{error.to_string()}</Error>),
+            )],
+        }
+    }
+}
+
+trait RangeExt {
+    fn to_text_range(self) -> TextRange;
+}
```
Comment (ematipico) [inline crates/biome_plugin_loader/src/analyzer_grit_plugin.rs:85]: Documentation is missing: the trait and the function.

File: crates/biome_plugin_loader/src/diagnostics.rs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.description(f)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Diagnostic)]
+#[diagnostic(
+	category = "plugin",
+    message = "Error compiling plugin",
+	severity = Error,
```
Comment (ematipico) [inline crates/biome_plugin_loader/src/diagnostics.rs:106]: nit: formatting

File: crates/biome_plugin_loader/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            analyzer_plugins: manifest
+                .rules
+                .into_iter()
+                .map(|rule| {
+                    if rule.as_os_str().as_encoded_bytes().ends_with(b".grit") {
+                        let plugin = AnalyzerGritPlugin::load(fs, &plugin_path.join(rule))?;
+                        Ok(Box::new(plugin) as Box<dyn AnalyzerPlugin>)
+                    } else {
+                        Err(PluginDiagnostic::unsupported_rule_format(markup!(
+                            "Unsupport rule format for plugin rule "
```
Comment (ematipico) [inline crates/biome_plugin_loader/src/lib.rs:79]:
```suggestion
                            "Unsupported rule format for plugin rule "
```

File: crates/biome_plugin_loader/src/snapshots/load_plugin_with_wrong_version.snap [hunk 5]
```diff
@@ -0,0 +1,10 @@
+---
+source: crates/biome_plugin_loader/src/lib.rs
+expression: content
+---
+plugin ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
+
+  × Cannot load plugin manifest
+    
+    Caused by:
+      version must be 1
```
Comment (ematipico) [inline crates/biome_plugin_loader/src/snapshots/load_plugin_with_wrong_version.snap:10]: I think these snapshots could be improved by providing the input/files provided.

**Review by ematipico — CHANGES_REQUESTED** [review]
I think the PR should be rebased because there are changes that don't belong here

File: crates/biome_plugin_loader/Cargo.toml (hunk 1 above)
Comment (arendjr) [reply crates/biome_plugin_loader/Cargo.toml:2]: Done

**Review by ematipico — APPROVED** [review]

**Pull request merged**